Update eol_data for API changes

samford opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

The endoflife.date API has been updated, so this modifies the URL in `SharedAudits.eol_data` to use the up to date URL and modifies the related logic in `FormulaAuditor.audit_eol` to work with the new response format. Specifically, there is now an `isEol` boolean value and the EOL date is found in `eolFrom`.

One wrinkle of the new setup is that 404 responses now return HTML content even if the request includes an `Accept: application/json` header. This handles these types of responses by catching `JSON::ParserError` but ideally we would parse the response headers and use `Utils::Curl.http_status_ok?` to check for a good response status before trying to parse the response body as JSON. [Edit: I worked on this but hit a wall because the tests return an `undefined method 'parse_curl_response' for module Utils::Curl` error if `Utils::Curl.parse_curl_output` is used in `eol_data`.]

I'll be gone for the next few hours but feel free to make any necessary changes to this to move it forward in the interim time. Otherwise, I'll come back to this later today.

daeho-ro commented on Library/Homebrew/utils/shared_audits.rb:
```diff
@@ -12,9 +12,17 @@ module SharedAudits
+      if result.status.success?
+        begin
+          json = JSON.parse(result.stdout)
+          json = nil if json&.dig("message")&.include?("Product not found")
```
This can be never happen. They removed 404.json and now it will not be json.

- https://github.com/endoflife-date/endoflife.date/pull/2080/files#diff-6e6db31acc12e2feb1ce488d68d8ad68a0b80e86005b2f47aeb20d97ea3eb41b

samford replied:
Thanks for confirming. I removed this in the latest push.

SMillerDev commented on Library/Homebrew/utils/shared_audits.rb:
```diff
@@ -12,10 +12,15 @@ module SharedAudits
-      json
+      result = Utils::Curl.curl_output("--location", "https://endoflife.date/api/v1/products/#{product}/releases/#{cycle}")
+
+      if result.status.success?
+        begin
+          JSON.parse(result.stdout)
+        rescue JSON::ParserError
+          nil
+        end
+      end
```
This might be easier to read 
```suggestion
      return unless result.status.success?

      json = begin
        JSON.parse(result.stdout)
      rescue JSON::ParserError
        nil
      end
```

samford replied:
That's what I had originally but unfortunately it doesn't pass `brew style` because RuboCop mistakenly thinks the return is occurring in a begin...end block.

botantony commented on Library/Homebrew/formula_auditor.rb:
```diff
@@ -612,11 +612,10 @@ def audit_eol
       metadata ||= SharedAudits.eol_data(name, formula.version.major_minor.to_s)
 
-      return if metadata.blank? || (eol = metadata["eol"]).blank?
+      return if metadata.blank? || (metadata.dig("result", "isEol") != true)
```
Should it be `isEol` field? I think `isMaintained` fits here better. F.e. `isEol` for Python version 3.10 is `false`, but we still want to update it (it receives security patches). `eolFrom` shows when `isMaintained` will become `false`, not `isEol`

daeho-ro commented on Library/Homebrew/utils/shared_audits.rb:
```diff
@@ -12,10 +12,15 @@ module SharedAudits
-      json = nil if json&.dig("message")&.include?("Product not found")
-      json
+      result = Utils::Curl.curl_output("--location", "https://endoflife.date/api/v1/products/#{product}/releases/#{cycle}")
+
+      if result.status.success?
+        begin
+          JSON.parse(result.stdout)
+        rescue JSON::ParserError
+          nil
+        end
+      end
```
What about this?
```suggestion
      JSON.parse(result.stdout) if result.status.success?
    rescue JSON::ParserError
      nil
```

samford replied:
That's technically possible (since this is already in a `begin` block) but I'm not sure that we need to be using a `begin` block for more than just the `JSON.parse` call. I refactored the method to reduce the `begin` block to only the `JSON.parse` call, allowing us to use `return unless result.status.success?` instead.

samford commented:
Thanks for the tests, @daeho-ro. I updated them a bit to make the JSON text more generic (rather than real nodejs data) and to add safe navigation operators to the `#dig` calls on nilable return values.

I refactored `eol_data` in the latest push to reduce the `begin` block to only the `JSON.parse` call, which allows for nicer code. Besides that, I added a check to return an existing value if the key is set (preventing duplicate fetches that `||=` would allow) and this will only set an `@eol_data` value if the `curl` exit status is successful (allowing a URL to be fetched again if previous attempts failed).

Bo98 reviewed (approved):
Thanks!
